core/include/detray/tools/cuboid_portal_generator.hpp

```diff
@@ -0,0 +1,231 @@
+ *
+ * (c) 2023 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s)
+#include "detray/definitions/geometry.hpp"
+#include "detray/definitions/indexing.hpp"
+#include "detray/definitions/qualifiers.hpp"
+#include "detray/masks/cuboid3D.hpp"
+#include "detray/tools/bounding_volume.hpp"
+#include "detray/tools/surface_factory_interface.hpp"
+
+// System include(s)
+#include <cassert>
+#include <limits>
+
+namespace detray {
+
+/// @brief Generates a portal box around a volume that already contains surfaces
+///
+/// @tparam detector_t the type of detector the volume belongs to.
+template <typename detector_t>
+class cuboid_portal_generator final
+    : public surface_factory_interface<detector_t> {
+
+    using scalar_t = typename detector_t::scalar_type;
+    using transform3_t = typename detector_t::transform3;
+
+    /// A functor to construct global bounding boxes around masks
+    struct bounding_box_creator {
+
+        using aabb_t = axis_aligned_bounding_volume<cuboid3D<>>;
+
+        template <typename mask_group_t, typename index_t>
+        DETRAY_HOST_DEVICE inline void operator()(
+            const mask_group_t &mask_group, const index_t &index,
+            const scalar_t envelope, const transform3_t &trf,
+            std::vector<aabb_t> &boxes) const {
+            // Local minimum bounding box
+            aabb_t box{mask_group[index], boxes.size(), envelope};
+            // Minimum bounding box in global coordinates
+            boxes.push_back(box.transform(trf));
+        }
+    };
+
+    public:
+    /// Use @param env as portal envelope
+    DETRAY_HOST
+    cuboid_portal_generator(const scalar_t env) : m_envelope{env} {}
+
+    /// @returns the id for the surface type (portal surfaces)
+    DETRAY_HOST
+    auto surface_type() const -> surface_id override {
+        return surface_id::e_portal;
+    }
+
+    /// @returns the number of rectangle portals this factory will produce
+    DETRAY_HOST
+    auto size() const -> dindex override { return 6u; }
+
+    DETRAY_HOST
+    void push_back(surface_data<detector_t> &&) override { /*Do nothing*/
+    }
+    DETRAY_HOST
+    auto push_back(std::vector<surface_data<detector_t>> &&)
+        -> void override { /*Do nothing*/
+    }
+
+    /// Create minimum aabbs around all module surfaces in the volume and then
+    /// construct world portals from the global bounding box.
+    /// This assumes that the module surfaces are already built inside the
+    /// argument containers.
+    ///
+    /// @param volume the volume the portals need to be added to.
+    /// @param surfaces the surface collection to wrap and to add the portals to
+    /// @param transforms the transforms of the surfaces.
+    /// @param masks the masks of the surfaces.
+    /// @param ctx the geometry context (not needed for portals).
+    DETRAY_HOST
+    auto operator()(typename detector_t::volume_type &volume,
```

Comment:
I just copied the portal creation from the telescope here, to make it accessible in the detector building example